### ISSUE TITLE
Alogp descriptor doesn't show found fragments for implicit hydrogen

### ISSUE THREAD

**Issue by marcabrus** [cdk/cdk#538]
I'm looking into differences of calculations for several chemistry software, in particular Logp.
Since Alogp is a common Qsar model based on atom center fragments contributions like described in [A.K. Ghose and G.M. Crippen, J. Comput. Chem. 1986], the calculation should be identical among several descriptors engine, unless some fragments are recognized differently among the several software.
In order to check if the recognized fragments are the same, I used the "alogpfrag" stored in the AlogpDescriptor object, that ideally should give the fragment label to each atom contained in the molecule (even hydrogen). As stated in the code, since hydrogen assignments are not cached, they disappear after Alogp calculation.
Since it could be a important thing to know, I would propose to change to public also the array of int "frags" stored in the descriptors.
I wrote a code that implements all the fragments stored in AlogP as a new block of descriptors (like in commercial software DRAGON) that let to evaluate easily possible bugs in fragment interpretation:

```java
import org.openscience.cdk.DefaultChemObjectBuilder;
import org.openscience.cdk.IImplementationSpecification;
import org.openscience.cdk.exception.CDKException;
import org.openscience.cdk.exception.InvalidSmilesException;
import org.openscience.cdk.interfaces.IAtomContainer;
import org.openscience.cdk.qsar.AbstractMolecularDescriptor;
import org.openscience.cdk.qsar.DescriptorSpecification;
import org.openscience.cdk.qsar.DescriptorValue;
import org.openscience.cdk.qsar.IMolecularDescriptor;
import org.openscience.cdk.qsar.result.DoubleArrayResult;
import org.openscience.cdk.qsar.result.DoubleArrayResultType;
import org.openscience.cdk.qsar.result.IDescriptorResult;
import org.openscience.cdk.smiles.SmilesParser;
import org.openscience.cdk.tools.CDKHydrogenAdder;
import org.openscience.cdk.tools.manipulator.AtomContainerManipulator;

public class AtomCenteredFragments extends AbstractMolecularDescriptor implements IMolecularDescriptor {

    static final String[] ACFNames = new String[121];
    static {
        // fragment names
        ACFNames[0] = "U-000";
        ACFNames[1] = "C-001";
        ACFNames[2] = "C-002";
        ACFNames[3] = "C-003";
        ACFNames[4] = "C-004";
        ACFNames[5] = "C-005";
        ACFNames[6] = "C-006";
        ACFNames[7] = "C-007";
        ACFNames[8] = "C-008";
        ACFNames[9] = "C-009";
        ACFNames[10] = "C-010";
        ACFNames[11] = "C-011";
        ACFNames[12] = "C-012";
        ACFNames[13] = "C-013";
        ACFNames[14] = "C-014";
        ACFNames[15] = "C-015";
        ACFNames[16] = "C-016";
        ACFNames[17] = "C-017";
        ACFNames[18] = "C-018";
        ACFNames[19] = "C-019";
        ACFNames[20] = "C-020";
        ACFNames[21] = "C-021";
        ACFNames[22] = "C-022";
        ACFNames[23] = "C-023";
        ACFNames[24] = "C-024";
        ACFNames[25] = "C-025";
        ACFNames[26] = "C-026";
        ACFNames[27] = "C-027";
        ACFNames[28] = "C-028";
        ACFNames[29] = "C-029";
        ACFNames[30] = "C-030";
        ACFNames[31] = "C-031";
        ACFNames[32] = "C-032";
        ACFNames[33] = "C-033";
        ACFNames[34] = "C-034";
        ACFNames[35] = "C-035";
        ACFNames[36] = "C-036";
        ACFNames[37] = "C-037";
        ACFNames[38] = "C-038";
        ACFNames[39] = "C-039";
        ACFNames[40] = "C-040";
        ACFNames[41] = "C-041";
        ACFNames[42] = "C-042";
        ACFNames[43] = "C-043";
        ACFNames[44] = "C-044";
        ACFNames[45] = "U-045";
        ACFNames[46] = "H-046";
        ACFNames[47] = "H-047";
        ACFNames[48] = "H-048";
        ACFNames[49] = "H-049";
        ACFNames[50] = "H-050";
        ACFNames[51] = "H-051";
        ACFNames[52] = "H-052";
        ACFNames[53] = "H-053";
        ACFNames[54] = "H-054";
        ACFNames[55] = "H-055";
        ACFNames[56] = "O-056";
        ACFNames[57] = "O-057";
        ACFNames[58] = "O-058";
        ACFNames[59] = "O-059";
        ACFNames[60] = "O-060";
        ACFNames[61] = "O-061";
        ACFNames[62] = "O-062";
        ACFNames[63] = "O-063";
        ACFNames[64] = "Se-064";
        ACFNames[65] = "Se-065";
        ACFNames[66] = "N-066";
        ACFNames[67] = "N-067";
        ACFNames[68] = "N-068";
        ACFNames[69] = "N-069";
        ACFNames[70] = "N-070";
        ACFNames[71] = "N-071";
        ACFNames[72] = "N-072";
        ACFNames[73] = "N-073";
        ACFNames[74] = "N-074";
        ACFNames[75] = "N-075";
        ACFNames[76] = "N-076";
        ACFNames[77] = "N-077";
        ACFNames[78] = "N-078";
        ACFNames[79] = "N-079";
        ACFNames[80] = "U-080";
        ACFNames[81] = "F-081";
        ACFNames[82] = "F-082";
        ACFNames[83] = "F-083";
        ACFNames[84] = "F-084";
        ACFNames[85] = "F-085";
        ACFNames[86] = "Cl-086";
        ACFNames[87] = "Cl-087";
        ACFNames[88] = "Cl-088";
        ACFNames[89] = "Cl-089";
        ACFNames[90] = "Cl-090";
        ACFNames[91] = "Br-091";
        ACFNames[92] = "Br-092";
        ACFNames[93] = "Br-093";
        ACFNames[94] = "Br-094";
        ACFNames[95] = "Br-095";
        ACFNames[96] = "I-096";
        ACFNames[97] = "I-097";
        ACFNames[98] = "I-098";
        ACFNames[99] = "I-099";
        ACFNames[100] = "I-100";
        ACFNames[101] = "F-101";
        ACFNames[102] = "Cl-102";
        ACFNames[103] = "Br-103";
        ACFNames[104] = "I-104";
        ACFNames[105] = "U-105";
        ACFNames[106] = "S-106";
        ACFNames[107] = "S-107";
        ACFNames[108] = "S-108";
        ACFNames[109] = "S-109";
        ACFNames[110] = "S-110";
        ACFNames[111] = "Si-111";
        ACFNames[112] = "B-112";
        ACFNames[113] = "U-113";
        ACFNames[114] = "U-114";
        ACFNames[115] = "P-115";
        ACFNames[116] = "P-116";
        ACFNames[117] = "P-117";
        ACFNames[118] = "P-118";
        ACFNames[119] = "P-119";
        ACFNames[120] = "P-120";

    }

    @Override
    public DescriptorValue calculate(IAtomContainer atomcontainer) {
        // TODO Auto-generated method stub
        //mock alogp descriptor, cloned to have frags public
        ACFCDK alogp = null;
        try {
            alogp = new ACFCDK();
            alogp.calculate(atomcontainer);
            DoubleArrayResult results = new DoubleArrayResult();
            for (int i = 0; i <= 120; i++) {
                results.add(alogp.frags[i]);
            }
            return new DescriptorValue(getSpecification(), getParameterNames(), getParameters(), results,
                getDescriptorNames());

        } catch (CDKException e) {
            return getDummyDescriptorValue(new CDKException("Error during atom typing" + e.getMessage()));

        }

    }
    private DescriptorValue getDummyDescriptorValue(Exception e) {
        DoubleArrayResult results = new DoubleArrayResult();
        for (int i = 0; i <= 120; i++) {
            results.add(Double.NaN);
        }
        return new DescriptorValue(getSpecification(), getParameterNames(), getParameters(), results,
            getDescriptorNames(), e);
    }

    @Override
    public IDescriptorResult getDescriptorResultType() {
        // TODO check if it's better integer result type, even if it can't be added NaN as dummy result
        return new DoubleArrayResultType(121);
    }

    @Override
    public String[] getDescriptorNames() {
        // Return names of descriptors

        return ACFNames;
    }

    @Override
    public String[] getParameterNames() {
        // TODO Auto-generated method stub
        return new String[0];
    }

    @Override
    public Object getParameterType(String arg0) {
        return null;
    }

    @Override
    public Object[] getParameters() {
        return null;
    }

    @Override
    public DescriptorSpecification getSpecification() {
        // added same specs of AlogP
        return new DescriptorSpecification("http://www.blueobelisk.org/ontologies/chemoinformatics-algorithms/#ALOGP",
            this.getClass().getName(), "The Chemistry Development Kit");
    }

    @Override
    public void setParameters(Object[] arg0) throws CDKException {

    }
}
```
@rajarshi  if you think it's useful It can be added among CDK descriptors as well

**Comment by marcabrus** [comment]
> Bit confused... I do know the AlogP is broken but I'm not quite sure what you're trying to convey here.

As you know, Alogp is one of the Dragon-like descriptors thaa Todd Martin implemented when i developed T.E.S.T..
I was curious that his implementation give very different values than Dragon (or VEGA -  implemented in our lab) software. The most useful thing of DRAGON's Alogp that implemented the fragments  used for calculation as molecular descriptors. Basically if you use that descriptors and the table from the publication you are able to replicate Alogp results. Currently this is impossible to do with CDK version of Alogp. The current version doesn't let you know what coefficients are assigned to hydrogens, since they are not cached. I think that the possibility to use fragments from Alogp calculation as descriptors can be useful to evaluate the difference between the interpretation for certain fragments.
Let me cite Jeffrey Plante in Journal of Cheminformatics 2018 10:61 https://doi.org/10.1186/s13321-018-0316-5. in his comparison of several alogp:
> All of the models perform better than the average of the
values in the dataset, herein designated the Arithmetic
Average Mean (AAM), with the sole exception of AlogP
from CDK, as implemented in KNIME. As this is so different from the AlogP implementation in VEGA there
are suspicions that the poor results are due to either an
error in the implementation or a user error. Trying varying methods of standardisation, i.e. all combinations of
aromatisation/dearomitisation and with explicit/implicit
hydrogen treatment did not improve the results so they
are left more as a curiosity than to be considered a proper
result.


I think that  if we know exactly which are the atoms that are classified in different way than the original publication, would be a good step to improve CDK AlogP predictions.
Moreover I thought that having 120 more molecular descriptors would be useful.

**Comment by johnmay** [comment]
I've looked at the code and don't see the issue you're having.... at least that comment about hydrogen's not being cached is redundant as you can just make the hydrogens explicit before you call it. The ```frags[]``` is already package-private so you can access it from outside og the class (providing you call it from the same package) so it's not require to make it public.

Are the fragment names "official" I'm certainly happy to add a convince method that provides that.

**Comment by johnmay** [comment]
Okay I see what you mean about the "H caching" although it's not anything to do with that. The issue is the htypes are set from adjacent atoms and so the ``alogpfrag[]`` is not updated. 

Technically this should work:

```java
public static void main(String[] args) throws CDKException {
        SmilesParser smipar = new SmilesParser(SilentChemObjectBuilder.getInstance());
        IAtomContainer mol = smipar.parseSmiles("Oc1c(Br)cccc1");
        AtomContainerManipulator.convertImplicitToExplicitHydrogens(mol);
        ALOGPDescriptor alogp = new ALOGPDescriptor();
        alogp.calculate(mol);
        for (int i = 0; i < mol.getAtomCount(); i++) {
            System.out.printf("%s-%03d\n", mol.getAtom(i).getSymbol(), alogp.alogpfrag[i]);
        }
    }
```
But you get:
```
O-057
C-034
C-034
Br-094
C-024
C-024
C-024
C-024
H-000
H-000
H-000
H-000
H-000
```

I'll see if I can go through and add in the relevant lines:

```
 frags[htype]++;
alogpfrag[((IAtom) ca.get(j)).getIndex()] = htype; // this is needed
```

...Although I might just rewrite the entire thing. The more I like the more I don't like it, particularly the use of the EState atom types.

**Comment by marcabrus** [comment]
the original table is the following:
![image](https://user-images.githubusercontent.com/335108/51828214-6e052280-22eb-11e9-81f1-8ad0ff414e4a.png)

there also examples for hydrogens are assigned differently from this table:
![image](https://user-images.githubusercontent.com/335108/51826682-ef5ab600-22e7-11e9-86c5-25f85ce3e9c4.png)

> R1 = H or CA3 where A = C or H, X = any electronegative atom, R3 = C or H or X. For atom types 53-55, the electronegative atom may be attached to the same or different carbon atom (as shown above).

As you can see the y use atom E-state to assign label to fragments. If you agree in the next days i will create an example dataset and try to check when there are mismatch in label assigning.

**Comment by johnmay** [comment]
Yep was skimming the paper last night. Sorry when I said EState I meant more that the atom type strings (SsssNH2) etc just a sign of the times I guess.

If you pull this commit - #539. I've made it so it will get the hydrogen types.

I think though I can see the issue already, running through ChEMBL get a lot of "U-000" for carbons and nitrogens. Will investigate (although really tempted to rewrite completely).

```java
import org.openscience.cdk.aromaticity.Aromaticity;
import org.openscience.cdk.aromaticity.ElectronDonation;
import org.openscience.cdk.exception.CDKException;
import org.openscience.cdk.graph.Cycles;
import org.openscience.cdk.interfaces.IAtomContainer;
import org.openscience.cdk.qsar.descriptors.molecular.ALOGPDescriptor;
import org.openscience.cdk.qsar.result.DoubleArrayResult;
import org.openscience.cdk.silent.SilentChemObjectBuilder;
import org.openscience.cdk.smiles.SmilesParser;
import org.openscience.cdk.tools.manipulator.AtomContainerManipulator;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;

public class ALogPDumper {

    private static final Aromaticity arom =
            new Aromaticity(ElectronDonation.daylight(),
                            Cycles.or(Cycles.all(), Cycles.all(6)));

    private static final SmilesParser smipar =
            new SmilesParser(SilentChemObjectBuilder.getInstance());

    private static final Charset utf8 = StandardCharsets.UTF_8;

    private static void prepare(IAtomContainer mol) {
        try {
            arom.apply(mol);
        } catch (CDKException e) {
            System.err.println("Error: Aromaticity could not be computed, " +
                               mol.getTitle() + ", " + e.getMessage());
        }
        AtomContainerManipulator.convertImplicitToExplicitHydrogens(mol);
    }

    public static void main(String[] args) throws CDKException {

        if (args.length < 1) {
            System.err.println("Usage: ALogpAtomTypes {input.smi}");
            return;
        }

        ALOGPDescriptor alogp = new ALOGPDescriptor();

        try (InputStream in = new FileInputStream(args[0]);
             InputStreamReader rdr = new InputStreamReader(in, utf8);
             BufferedReader brdr = new BufferedReader(rdr)) {
            String line;
            while ((line = brdr.readLine()) != null) {
                IAtomContainer mol;
                try {
                    mol = smipar.parseSmiles(line);
                } catch (CDKException ex) {
                    System.err.println("Error: Bad SMILES, " + line);
                    continue;
                }
                prepare(mol);
                DoubleArrayResult result = (DoubleArrayResult) alogp.calculate(mol).getValue();
                if (Double.isNaN(result.get(0))) {
                    System.err.println("Error: Could not compute ALogP" + line);
                    continue;
                }
                System.out.printf("%s AlogP=%.3f AlogP2=%.3f amr=%.3f\n",
                                  mol.getTitle(),
                                  result.get(0),
                                  result.get(1),
                                  result.get(2));
                for (int i = 0; i < mol.getAtomCount(); i++) {
                    System.out.printf("[%d] := %s-%03d\n",
                                      i,
                                      mol.getAtom(i).getSymbol(),
                                      alogp.alogpfrag[i]);
                }
            }
        } catch (IOException e) {
            System.err.println("Error: IO Error, " + e.getMessage());
        }
    }
}
```

**Comment by johnmay** [comment]
Do you know what ``R-N=X`` should be assigned to?

**Comment by marcabrus** [comment]
Looking at definitions:

R represents any group linked through carbon;
X represents any electronegative atom (O, N, S, P, Se, halogens);
so i think is a N double bonded with an heteroatom and single bonded with a r group.
Todeschini's book  it has been changed to 

78 | N-079 | N+ (positively charged)
I need to check for the reason (maybe a more recent publication)

**Comment by johnmay** [comment]
Actually we have multiple aromaticity models so it should be possible to exactly match the original.

**Comment by marcabrus** [comment]
> Depends on aromatic model I guess.. so either of the nitrogens here:
> 
> ![image](https://user-images.githubusercontent.com/983232/51833638-ea016980-22ef-11e9-8ca0-97d10d0a9d82.png)

I need to correct myself the paper says:

> Aromatic bond is expressed as - - as in benzene or delocalized bonds such as the N-O bond in a nitro group; ... represents aromatic single bonds as the C-N bond in pyrrole. 


this image could explain the interpretation:
![image](https://user-images.githubusercontent.com/335108/51841108-1f1cb480-230e-11e9-8c94-70ba4f683f3f.png)
as we can see the N marked by 1 are pyridine like nitrogen, and so:
> the C- -N bond order in pyridine may be considered as 2 while we have one such bond and 1.5 when we have two such bonds;

If I'm right I guess the correct label for the two nitrogen is R- -N- -X (N-075)

**Comment by marcabrus** [comment]
I need to say that Alberto Manganaro from Kode developed our AlogP version. This AlogP version is more accurate and it has the correct definitions of Atom Centered Fragments. Indeed based on a modified version of cdk 1.4.9 so it's not very simple to work on it. Maybe I could ask him if he agrees to pass  the source code and we could check how to merge it ( it's not based on E-state descriptors). In the meantime in my repository there is a KNIME workflow that calculates also Atom Centred Fragments in VEGA version, among other descriptors. it's useful for check the differences in calculation. If you are not familiar with KNIME I can provide a command line version.
My plans were to slowly pass all the descriptors calculated in VEGA into CDK. I think I started with the wrong block.

**Comment by johnmay** [comment]
I do know KNIME but not a fan and would prefer command line version. Out of interest do you know what the modifications to 1.4.9? Very old version I think even before I started contributing.

I had a neat idea actually earlier that should allow simple but extremely fast atom typing so I might just work from the paper and use the VEGA atom types to make sure things line up.

**Comment by johnmay** [comment]
Is it not possible just to get the VEGA jar/code anymore..? I followed the links from VEGA HUB to https://chm.kode-solutions.net/products_vega.php which say you can download it from www.vega-qsar.eu -> which redirect to VEGA Hub.

**Comment by marcabrus** [comment]
> Is it not possible just to get the VEGA jar/code anymore..? I followed the links from VEGA HUB to https://chm.kode-solutions.net/products_vega.php which say you can download it from [www.vega-qsar.eu](http://www.vega-qsar.eu) -> which redirect to VEGA Hub.

You need to register in order to download VEGA. It's free, but it's to keep track of downloads.
this is  the command line version. very old and maybe not completely tested. I made it almost one year ago but since then i used only KNIME version because it's mode useful in workflows. the bat cang give you the example for using it. tomorrow I'll provide at least a guide

**Comment by marcabrus** [comment]
> I do know KNIME but not a fan and would prefer command line version. Out of interest do you know what the modifications to 1.4.9? Very old version I think even before I started contributing.

from the original atomcontainer it has been created a new class, insilicomolecule, that contains different objects, besides different methods. The methods underlying this class depend on often deprecated cdk methods. One for all aromaticity ( the default model i thing has changed, so we need to specify it). there is also a custom atom numbering and canonicalization, nitro group standardization etc. I don't know every detail because It has been done before I came to the lab.

**Comment by johnmay** [comment]
> One for all aromaticity ( the default model i thing has changed, so we need to specify it).

There is no default model anymore so saying it changed is a bit wrong. You can a) still call the old by including cdk-legacy module, b) exactly replicate the old one (but it will run faster):

```java
Aromaticity arom = new Aromaticity(ElectronDontation.cdk(), Cycles.cdkLegacy());
```

The key point is the model of aromaticity is what atoms donate how many electrons and which rings do you test **4n+2** on, for example I can still use the CDK AtomType based model (not my fav) but apply it to a different set of rings.

```java
Aromaticity arom = new Aromaticity(ElectronDontation.cdk(), Cycles.or(Cycles.all(),Cycles.all(6)));
```

**Comment by johnmay** [comment]
Also thanks for ZIP will get that tested. I think from my first pass there were a lot of simple obvious things that needed fixing - but ultimately I think it's possible to go much faster 500K molecules per second with a rewrite.

**Comment by marcabrus** [comment]
> > One for all aromaticity ( the default model i thing has changed, so we need to specify it).
> 
> There is no default model anymore so saying it changed is a bit wrong. You can a) still call the old by including cdk-legacy module, b) exactly replicate the old one (but it will run faster):
> 
> ```java
> Aromaticity arom = new Aromaticity(ElectronDontation.cdk(), Cycles.cdkLegacy());
> ```
> 
> The key point is the model of aromaticity is what atoms donate how many electrons and which rings do you test **4n+2** on, for example I can still use the CDK AtomType based model (not my fav) but apply it to a different set of rings.
> 
> ```java
> Aromaticity arom = new Aromaticity(ElectronDontation.cdk(), Cycles.or(Cycles.all(),Cycles.all(6)));
> ```

I think i need to get deeper in this, because a the moment we consider only huekel rule, and ring complexes or rings with extenal double bonds are excluded. Is there a wiki?

**Comment by marcabrus** [comment]
> Also thanks for ZIP will get that tested. I think from my first pass there were a lot of simple obvious things that needed fixing - but ultimately I think it's possible to go much faster 500K molecules per second with a rewrite.

I know i'ts very raw, I did't create a job method, and neither a support for multicore, or the possibility to choose descriptors. Hope I will do it someday.

**Comment by johnmay** [comment]
Nah it's fine got it working. I did forgot just how slow CDK 1.4 was though...

**Comment by johnmay** [comment]
So found some simple things in the CDK that need fixing. I can't work out if there is a way to get the atom type assignment from the Descriptors.jar you shared. I computed the LogPs for ChEMBL and then have been looking at differences in detail. I can get the fragment values as follows but in this particular case I was testing I think it's missed one of the nitrogens. Getting the per atom assigned (as you suggested for CDK) would be useful.

```java
SmilesParser smipar = new SmilesParser(SilentChemObjectBuilder.getInstance());
    InsilicoMolecule m = SmilesMolecule.Convert("Brc1cccc(Nc2ncnc3ccncc23)c1NCCN4CCOCC4");
    ALogP alogp = new ALogP();
    AtomCenteredFragments frags = new AtomCenteredFragments();
    frags.Calculate(m);
    double[] frags2 = frags.GetAllValues();
    double sum = 0;
    for (int i = 0; i < frags2.length; i++) {
      if (frags2[i] > 0) {
        System.out.println(i + "\t" + (int) frags2[i]);
        sum += frags2[i];
      }
    }
    System.out.println((int) sum + " atoms");
    alogp.Calculate(m);
    System.out.println(alogp.GetByName("ALogP").getFormattedValue());
```

Output:
```
6	6
24	4
25	1
26	3
27	2
28	1
29	1
30	1
47	16
49	3
50	2
59	1
68	1 // Nx1
73	1 // Nx1
75	3 // Nx3
94	1
47 atoms // should be 48 atoms!
```

**Comment by johnmay** [comment]
And another one... ``COc1c(O)cc(O)c(C(=N)Cc2ccc(O)cc2)c1O``  the nitrogen atom type is missed, I get 74 for ``=NH`` in CDK not sure if this is right but looks reasonable.

**Comment by marcabrus** [comment]
> So found some simple things in the CDK that need fixing. I can't work out if there is a way to get the atom type assignment from the Descriptors.jar you shared. I computed the LogPs for ChEMBL and then have been looking at differences in detail. I can get the fragment values as follows but in this particular case I was testing I think it's missed one of the nitrogens. Getting the per atom assigned (as you suggested for CDK) would be useful.
> 
> ```java
> SmilesParser smipar = new SmilesParser(SilentChemObjectBuilder.getInstance());
>     InsilicoMolecule m = SmilesMolecule.Convert("Brc1cccc(Nc2ncnc3ccncc23)c1NCCN4CCOCC4");
>     ALogP alogp = new ALogP();
>     AtomCenteredFragments frags = new AtomCenteredFragments();
>     frags.Calculate(m);
>     double[] frags2 = frags.GetAllValues();
>     double sum = 0;
>     for (int i = 0; i < frags2.length; i++) {
>       if (frags2[i] > 0) {
>         System.out.println(i + "\t" + (int) frags2[i]);
>         sum += frags2[i];
>       }
>     }
>     System.out.println((int) sum + " atoms");
>     alogp.Calculate(m);
>     System.out.println(alogp.GetByName("ALogP").getFormattedValue());
> ```
> 
> Output:
> 
> ```
> 6	6
> 24	4
> 25	1
> 26	3
> 27	2
> 28	1
> 29	1
> 30	1
> 47	16
> 49	3
> 50	2
> 59	1
> 68	1 // Nx1
> 73	1 // Nx1
> 75	3 // Nx3
> 94	1
> 47 atoms // should be 48 atoms!
> ```

I think i need to pass you source code, at least of for atom centered fragments ad alogp classes. Do you have a email to send you the two classes? lookin at the code you wrote i think that for one fragment has been assigned -999. Try to add another if.

**Comment by johnmay** [comment]
> i think that for one fragment has been assigned -999

Checked that and nope. I have the source code the JAR you sent.

**Comment by marcabrus** [comment]
> And another one... `COc1c(O)cc(O)c(C(=N)Cc2ccc(O)cc2)c1O` the nitrogen atom type is missed, I get 74 for `=NH` in CDK not sure if this is right but looks reasonable.

sorry I replied after a while, dragon assings n-70 to thi fragment (so i think this is a error in vega porting) but I need some time to understand which atom is it

**Comment by johnmay** [comment]
That's fine just good to know, I've managed to the total delta down on the first 1K compounds from ~8000 to ~4000. Best summed up as "death by 1000 paper cuts", there isn't one major thing wrong just lots of small things. I think I will rewrite but want to get as close as possible with the current code first.

**Comment by marcabrus** [comment]
How many compounds are you testing right now?

**Comment by johnmay** [comment]
That was 1000K from ChEMBL but i've ported the ``AtomCenteredFragments`` to CDK 2.3-SNAPSHOT so I can test a lot faster, i.e. ~2 million. One tricky point is it looks like it needs implicit hydrogens.

**Comment by johnmay** [comment]
I can now rapidly produce a SMILES like this:

```Cc1cccc(c1)N2CCN(CCCON3C(=O)c4ccccc4C3=O)CC2 |$_AV:;;;;;;;71&#47;72;;;68&#47;72$|```

which i can view:

[https://www.simolecule.com/cdkdepict/depict/bow/svg?smi=Cc1cccc(c1)N2CCN(CCCON3C(%3DO)c4ccccc4C3%3DO)CC2%20%7C%24_AV%3A%3B%3B%3B%3B%3B%3B%3B71%26%2347%3B72%3B%3B%3B68%26%2347%3B72%24%7C&abbr=on&hdisp=bridgehead&showtitle=false&zoom=1.6&annotate=atomvalue](https://www.simolecule.com/cdkdepict/depict/bow/svg?smi=Cc1cccc(c1)N2CCN(CCCON3C(%3DO)c4ccccc4C3%3DO)CC2%20%7C%24_AV%3A%3B%3B%3B%3B%3B%3B%3B71%26%2347%3B72%3B%3B%3B68%26%2347%3B72%24%7C&abbr=on&hdisp=bridgehead&showtitle=false&zoom=1.6&annotate=atomvalue)

![](https://www.simolecule.com/cdkdepict/depict/bow/svg?smi=Cc1cccc(c1)N2CCN(CCCON3C(%3DO)c4ccccc4C3%3DO)CC2%20%7C%24_AV%3A%3B%3B%3B%3B%3B%3B%3B71%26%2347%3B72%3B%3B%3B68%26%2347%3B72%24%7C&abbr=on&hdisp=bridgehead&showtitle=false&zoom=1.6&annotate=atomvalue)

In this case CDK says 68 and 71 when they should both be 72 (VEGA).

**Comment by marcabrus** [comment]
> I can now rapidly produce a SMILES like this:
> 
> `Cc1cccc(c1)N2CCN(CCCON3C(=O)c4ccccc4C3=O)CC2 |$_AV:;;;;;;;71&#47;72;;;68&#47;72$|`
> 
> which i can view:
> 
> https://www.simolecule.com/cdkdepict/depict/bow/svg?smi=Cc1cccc(c1)N2CCN(CCCON3C(%3DO)c4ccccc4C3%3DO)CC2%20%7C%24_AV%3A%3B%3B%3B%3B%3B%3B%3B71%26%2347%3B72%3B%3B%3B68%26%2347%3B72%24%7C&abbr=on&hdisp=bridgehead&showtitle=false&zoom=1.6&annotate=atomvalue
> 
> ![](https://camo.githubusercontent.com/d4900ce5932a1fab6940735dc634b3e3186323bc/68747470733a2f2f7777772e73696d6f6c6563756c652e636f6d2f63646b6465706963742f6465706963742f626f772f7376673f736d693d43633163636363286331294e3243434e284343434f4e3343282533444f29633463636363633443332533444f294343322532302537432532345f415625334125334225334225334225334225334225334225334237312532362532333437253342373225334225334225334236382532362532333437253342373225323425374326616262723d6f6e2668646973703d627269646765686561642673686f777469746c653d66616c7365267a6f6f6d3d312e3626616e6e6f746174653d61746f6d76616c7565)
> 
> In this case CDK says 68 and 71 when they should both be 72 (VEGA).

Ok this is beyond my imagination.
Do youn know how to interpret > or <? I don't find any clue. There is lot of space of interpretation looking at the table from the original publication(especially for S typing, it will be also a struggle). But in this case  I think that CDK intepretation is quite correct. 
N (68/72) has 3 aliphatic bonds, while N (71/72) has 2 aliphatic bonds and is attached to an aromatic ring. I' can't justify Vega atom typing:
```java
            // fragment with particular groups
            if (!AromPyrroleLike)
                if (((VD==3) && (sOCR==1)) ||   // >N-OCR
                    (((VD+nH)==3) && (sXdX>0)))     // >N-X=X  NOTE: also >N-R=X for Dragon compatibility
                    { FragCount[72]++; return; }
```
this seems to be correct, in need to check further

**Comment by johnmay** [comment]
In line notation ``>`` means two single bonds.

For example

``
>CH-OH Sp3
=CH-OH Sp2
``

**Comment by johnmay** [comment]
> Edit: my vega version gives the correct atom typing, This seems more difficult to understand now.

It could be the decompiler gets confused with variables. Are you able to share the ``AtomCenteredFragments.java`` to my email (will send)?

**Comment by johnmay** [comment]
Hmm can't figure out how to message, we do have a mailing list BTW (cdk-user) but my email is: john.wilksinsonmay@gmail.com

**Comment by marcabrus** [comment]
> > Edit: my vega version gives the correct atom typing, This seems more difficult to understand now.
> 
> It could be the decompiler gets confused with variables. Are you able to share the `AtomCenteredFragments.java` to my email (will send)?

I know sometimes jd-decompilers does that, Vega is opensource, we didn't publish yet on a repository but I of course I can share it.

**Comment by johnmay** [comment]
Thanks for code has made it much easier to identify actual differences. This is an interesting one, it's actually not clear to me which is correct as both valid. The definitions are:

```
R--CX--R 29
R--CR..X 34
```

It is no stipulated whether the bond to the CX/CR must be single or aromatic.

[![](https://www.simolecule.com/cdkdepict/depict/bow/svg?smi=CCOC(%3DO)c1cc2cc(ccc2%5BnH%5D1)C(%3DO)O%20%7C%24_AV%3A%3B%3B%3B%3B%3B%3B%3B%3B%3B%3B%3B%3B26%26%2347%3B34%24%7C%20CHEMBL503643&abbr=off&hdisp=bridgehead&showtitle=false&zoom=1.6&annotate=atomvalue)](https://www.simolecule.com/cdkdepict/depict/bow/svg?smi=CCOC(%3DO)c1cc2cc(ccc2%5BnH%5D1)C(%3DO)O%20%7C%24_AV%3A%3B%3B%3B%3B%3B%3B%3B%3B%3B%3B%3B%3B26%26%2347%3B34%24%7C%20CHEMBL503643&abbr=off&hdisp=bridgehead&showtitle=false&zoom=1.6&annotate=atomvalue)
```
CCOC(=O)c1cc2cc(ccc2[nH]1)C(=O)O |$_AV:;;;;;;;;;;;;26&#47;34$| CHEMBL503643
```

**Comment by marcabrus** [comment]
> CCOC(=O)c1cc2cc(ccc2[nH]1)C(=O)O

It's not the first time I found such a ambiguity. do you have any Idea how to find the original dataset? I don't think in 1998 they provide supplementary materials, but it would be the ideal benchmark dataset.

**Comment by johnmay** [comment]
Not that I can see, it is a curiosity but tautomers are still the big issue :-).

**Comment by marcabrus** [comment]
I know two tools, lychi https://github.com/ncats/lychi (based on Jchem free  libraries) written in java, developed by ncats and also molvs, written in python (rdkit bindings) that standardize also for tautomers. They are based on reaction libraries, easy to implement. Another work in progress for Vega

**Comment by johnmay** [comment]
This is one of the fastest Tautomer generators you can get: https://github.com/cdk/cdk/tree/sd-tautomer. It's not based on reactions :-) - which is why it's so fast. I explain in my thesis but not patched it full in yet.

The issue with LogP though isn't just generating the tautomers it's choosing/getting a consistent value: [relevant talk](https://youtu.be/zJhQaNKiP84?t=534)

**Comment by marcabrus** [comment]
> This is one of the fastest Tautomer generators you can get: https://github.com/cdk/cdk/tree/sd-tautomer. It's not based on reactions :-) - which is why it's so fast. I explain in my thesis but not patched it full in yet.
> 
> The issue with LogP though isn't just generating the tautomers it's choosing/getting a consistent value: [relevant talk](https://youtu.be/zJhQaNKiP84?t=534)

Do you propose take in account also other tautomers in logP calculation (i think ACD/see already does this)? or do you think to generate all tautomers and then find the most stable? In this case I think that pH could influence the stability of a tautomer. Do you have something to read about your method and how far you have gone?

**Comment by johnmay** [comment]
> Do you have something to read about your method and how far you have gone?

My method is just to do with the tautomer generation, I do not select a preferred one but select "a" canonical one see my [thesis](https://www.repository.cam.ac.uk/handle/1810/246652). I have only cared about LogP since this patch.

As per #540 I've improved things dramatically, still may be useful to rewrite but this was very useful. I'm left now with atom types that look either ambiguous or aren't assigned by VEGA.
